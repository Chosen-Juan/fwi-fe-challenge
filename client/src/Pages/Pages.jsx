import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchPlayers } from '../appState/actions';

import './Pages.scss';

export class Pages extends Component {
  static propTypes = {
    fetchPlayers: PropTypes.func.isRequired,
    from: PropTypes.number,
    size: PropTypes.number,
    total: PropTypes.number,
  };

  state = {
    total: 0,
    size: 25,
    from: 0,
    numberOfPages: 0,
    selectedPage: 0,
  };

  componentWillReceiveProps(props) {
    const { total, from } = props;
    this.setState({
      total, from,
      numberOfPages: Math.ceil(total/this.state.size),
      selectedPage: Math.ceil(from/this.state.size)
    });
  }


  changePage = page => {
    const { size } = this.state;
    const from = page * size;
    this.props.fetchPlayers(from, size);
  };

  nextPage = () => {
    const { selectedPage, numberOfPages } = this.state;
    if(selectedPage >= (numberOfPages - 1)) {
      return;
    }
    this.changePage(selectedPage + 1);
  };

  previousPage = () => {
    const { selectedPage } = this.state;
    if(selectedPage <= 0) {
      return;
    }
    this.changePage(selectedPage - 1);
  };

  getPages = () => {
    const { numberOfPages, selectedPage } = this.state;
    let pages = [];
    for(let x = 0; x < numberOfPages; x++) {
      pages.push(
        <button
          key={x}
          onClick={() => this.changePage(x)}
          className={selectedPage === x ? "button is-primary": "button is-primary is-outlined"}
        >
          {x + 1}
        </button>
      );
    }
    return pages;
  };

  render() {
    return (
      <div className="pages level is-mobile">
        <div className="level-left">
          <button id="button-previous" className="button" onClick={this.previousPage}>{'<'}</button>
        </div>
        <div className="level-item">
          {this.getPages()}
        </div>
        <div className="level-right">
          <button id="button-next" className="button" onClick={this.nextPage}>{'>'}</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = ({ players: { from, size, total} }) => ({
  from, size, total
});

const mapDispatchToProps = {
  fetchPlayers
};

export default connect(mapStateToProps, mapDispatchToProps)(Pages);
