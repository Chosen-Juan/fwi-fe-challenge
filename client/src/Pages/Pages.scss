.pages {
  margin-top: 1rem;
  margin-bottom: 1rem;

  .level-left {
    margin-left: 2rem;
  }

  .level-right {
    margin-right: 2rem;
  }
}