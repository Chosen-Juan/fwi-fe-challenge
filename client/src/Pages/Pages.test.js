import React from 'react';
import { shallow } from 'enzyme';

import { Pages } from './Pages';

describe('Pages', () => {
  const numberOfPages = 4;
  const fetchPlayersMock = jest.fn();
  const defaultProps = {
    fetchPlayers: fetchPlayersMock,
    from: 0,
    total: 100
  };
  const willReceivePropsSpy = jest.spyOn(Pages.prototype, 'componentWillReceiveProps');
  const pages = shallow(<Pages {...defaultProps} />);
  pages.setProps(defaultProps);

  it('should match snapshot', () => {
    expect(pages).toMatchSnapshot();
  });

  it('should have exact number of pages', () => {
    expect(willReceivePropsSpy).toHaveBeenCalledTimes(1);
    expect(pages.state('numberOfPages')).toEqual(numberOfPages);
  });

});
