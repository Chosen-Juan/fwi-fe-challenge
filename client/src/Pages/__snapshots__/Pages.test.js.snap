// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages should match snapshot 1`] = `
<div
  className="pages level is-mobile"
>
  <div
    className="level-left"
  >
    <button
      className="button"
      id="button-previous"
      onClick={[Function]}
    >
      &lt;
    </button>
  </div>
  <div
    className="level-item"
  >
    <button
      className="button is-primary"
      key="0"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="button is-primary is-outlined"
      key="1"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="button is-primary is-outlined"
      key="2"
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="button is-primary is-outlined"
      key="3"
      onClick={[Function]}
    >
      4
    </button>
  </div>
  <div
    className="level-right"
  >
    <button
      className="button"
      id="button-next"
      onClick={[Function]}
    >
      &gt;
    </button>
  </div>
</div>
`;
